File: 1.easy/0001. Two Sum.java
class Solution {

    public int[] twoSum(int[] nums, int target) {
        int[][] withIdx = addIndexesToArr(nums);
        Arrays.sort(withIdx, (a, b) -> Integer.compare(a[0], b[0]));
        int front = 0, back = nums.length - 1;
        while (front < back) {
            if (withIdx[front][0] + withIdx[back][0] == target) {
                return new int[] { withIdx[front][1], withIdx[back][1] };
            } else if (withIdx[front][0] + withIdx[back][0] <= target) {
                front++;
            } else back--;
        }
        return new int[] { -1 };
    }

    public int[][] addIndexesToArr(int[] nums) {
        int[][] result = new int[nums.length][2];
        for (int i = 0; i < nums.length; i++) {
            result[i][0] = nums[i];
            result[i][1] = i;
        }
        return result;
    }
}

File: 1.easy/0001. Two Sum.cpp
#include <vector>
#include <unordered_map>
#include <iostream>

using namespace std;

class Solution
{
public:
    vector<int> twoSum(vector<int> &nums, int target)
    {
        int n = nums.size();
        unordered_map<int, int> nums_map;
        for (int i = 0; i < n; i++)
        {
            nums_map[nums[i]] = i;
        }
        for (int i = 0; i < n; i++)
        {
            int complement = target - nums[i];
            if (nums_map.count(complement) && nums_map[complement] != i)
            {
                return {i, nums_map[complement]};
            }
        }
        return {};
    }
};

int main()
{
    Solution solution;
    vector<int> nums = {2, 7, 11, 15};
    int target = 9;
    vector<int> expected = {0, 1};
    vector<int> result = solution.twoSum(nums, target);
    cout << "Result: " << result[0] << ", " << result[1] << endl;
    cout << "Expected: " << expected[0] << ", " << expected[1] << endl;
    return 0;
}